files/deadsnakes.list

```diff
@@ -1,2 +1,5 @@
 deb [signed-by=/etc/apt/keyrings/deadsnakes.gpg] http://ppa.launchpad.net/deadsnakes/ppa/ubuntu focal main
 deb-src [signed-by=/etc/apt/keyrings/deadsnakes.gpg] http://ppa.launchpad.net/deadsnakes/ppa/ubuntu focal main
+
+deb [signed-by=/etc/apt/keyrings/deadsnakes.gpg] http://ppa.launchpad.net/deadsnakes/nightly/ubuntu focal main
+deb-src [signed-by=/etc/apt/keyrings/deadsnakes.gpg] http://ppa.launchpad.net/deadsnakes/nightly/ubuntu focal main
```

Comment:
Now that b1 is available on the regular PPA, we shouldn't need the nightly builds any more, unless there's something we want from b2 before it hits the regular PPA.

Reply:
I was assuming I'd be able to remove this and had for my local testing, but there are some things we'll need from b2. Running the sanity tests with 3.12.0b1 crashes with

<details open>
<summary>raise Exception('Pattern "%s" did not match values:\n%s' % (pattern, '\n'.join(unmatched)))</summary>
<br>

```bash
Installing requirements for Python 3.12 [venv]
Stream command with data: /root/.ansible/test/venv/sanity.pep8/3.12/5e7a1ff1/bin/python
Working directory: /root/ansible
Program found: /root/.ansible/test/venv/sanity.pep8/3.12/5e7a1ff1/bin/python
HOME=/root
LC_ALL=en_US.UTF-8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
Using cached pip 23.1.2 bootstrap script: /root/.ansible/test/cache/get_pip_23_1_2.py
Execute command: /root/.ansible/test/venv/sanity.pep8/3.12/5e7a1ff1/bin/python /tmp/ansible-test-onevc9f5-pip.py --disable-pip-version-check pip==23.1.2 setuptools==67.7.2 wheel==0.37.1
Collecting pip==23.1.2
  Using cached pip-23.1.2-py3-none-any.whl (2.1 MB)
Collecting setuptools==67.7.2
  Using cached setuptools-67.7.2-py3-none-any.whl (1.1 MB)
Collecting wheel==0.37.1
  Using cached wheel-0.37.1-py2.py3-none-any.whl (35 kB)
Installing collected packages: wheel, setuptools, pip
Successfully installed pip-23.1.2 setuptools-67.7.2 wheel-0.37.1
Execute command: /root/.ansible/test/venv/sanity.pep8/3.12/5e7a1ff1/bin/python /tmp/ansible-test-onevc9f5-pip.py install --disable-pip-version-check -r requirements/sanity.pep8.txt
Collecting pycodestyle==2.10.0 (from -r requirements/sanity.pep8.txt (line 2))
  Downloading pycodestyle-2.10.0-py2.py3-none-any.whl (41 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.3/41.3 kB 5.7 MB/s eta 0:00:00
Installing collected packages: pycodestyle
Successfully installed pycodestyle-2.10.0
Execute command: /root/.ansible/test/venv/sanity.pep8/3.12/5e7a1ff1/bin/python /tmp/ansible-test-onevc9f5-pip.py uninstall -y --disable-pip-version-check pip wheel
Command exited with status 0 after 5.970986604690552 seconds.
Run command: /root/.ansible/test/venv/sanity.pep8/3.12/5e7a1ff1/bin/python -m pycodestyle --max-line-length 160 --config /dev/null --ignore E203,E402,E741,W503,W504 .azure-pipelines/scripts/combine-coverage.py .azure-pipeli ...
Working directory: /root/ansible
Program found: /root/.ansible/test/venv/sanity.pep8/3.12/5e7a1ff1/bin/python
HOME=/root
LC_ALL=en_US.UTF-8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
Command exited with status 1 after 17.027377605438232 seconds.
Traceback (most recent call last):
  File "/root/ansible/bin/ansible-test", line 45, in <module>
    main()
  File "/root/ansible/bin/ansible-test", line 36, in main
    cli_main(args)
  File "/root/ansible/test/lib/ansible_test/_internal/__init__.py", line 84, in main
    args.func(config)
  File "/root/ansible/test/lib/ansible_test/_internal/commands/sanity/__init__.py", line 278, in command_sanity
    result = test.test(args, sanity_targets, virtualenv_python)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/ansible/test/lib/ansible_test/_internal/commands/sanity/pep8.py", line 92, in test
    results = parse_to_list_of_dict(pattern, stdout)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/ansible/test/lib/ansible_test/_internal/util.py", line 1019, in parse_to_list_of_dict
    raise Exception('Pattern "%s" did not match values:\n%s' % (pattern, '\n'.join(unmatched)))
Exception: Pattern "^(?P<path>[^:]*):(?P<line>[0-9]+):(?P<column>[0-9]+): (?P<code>[WE][0-9]{3}) (?P<message>.*)$" did not match values:
test/lib/ansible_test/_internal/commands/integration/__init__.py:873:-61: E231 missing whitespace after ':'
test/lib/ansible_test/_internal/commands/integration/__init__.py:873:-60: E221 multiple spaces before operator
test/lib/ansible_test/_internal/commands/integration/__init__.py:879:-145: E231 missing whitespace after ':'
test/lib/ansible_test/_internal/commands/integration/__init__.py:879:-144: E221 multiple spaces before operator
test/lib/ansible_test/_internal/commands/integration/__init__.py:879:-95: E221 multiple spaces before operator
test/lib/ansible_test/_internal/host_profiles.py:155:-108: E231 missing whitespace after ':'
test/lib/ansible_test/_internal/host_profiles.py:155:-107: E241 multiple spaces after ':'
test/lib/ansible_test/_internal/host_profiles.py:155:-19: E231 missing whitespace after ','
test/lib/ansible_test/_internal/host_profiles.py:155:-6: E231 missing whitespace after ','
```
</details>

This issue was fixed today: https://github.com/python/cpython/commit/103ae4e36e965b2d693c90bb6cb699f845903950.

I commented out that line in `test/lib/ansible_test/_internal/util.py` raising the exception to see the rest of the test run: https://gist.github.com/s-hertel/d95e907dedd9c2235c6bcc09c88dc649

The pep8 errors are due to either the fstring bug linked above or a tokenization change that was fixed for b2 https://github.com/python/cpython/commit/2c02c6886739f0ed420d900b2a29933bc1c5df37.

This leaves
* an astroid issue (maybe just needs to add support for https://github.com/python/cpython/commit/24d8b88420b81fc60aeb0cbcacef1e72d633824a).
* mypy issues (see mypy issue 15277 for python3.12 tracking, though our code also needs adjustments)
* missing libyaml support in PyYAML

Here's an updated test run, using an ansible-core-test-container built yesterday with Python 3.12 from the nightly deadsnakes repo, i.e. it contains the multiline string fix, but not the fstring fix that was merged today (which I'm still hacking around in `test/lib/ansible_test/_internal/util.py` to see the rest of the tests): https://gist.github.com/s-hertel/7d05604e245af23a645bfa938d53fb6c.

I'm currently looking at the astroid issue, but trying to get the pylint test suite passing has led to a number of argparse/ast/etc rabbitholes without much progress.

Reply:
@s-hertel For [sanity test requirements updates](https://github.com/ansible/ansible/issues/80429), such as for `mypy` and `pyyaml`, coordinate with @Akasurde.